cascade-prelude/src/Cascade/Log/Message.hs

```diff
@@ -0,0 +1,121 @@
+  , time     :: Time
+  , location :: CallStack
+  }
+
+log :: WithLog env Message m => MonadIO m => Scope -> Severity -> Text -> m ()
+log scope severity message = do
+  time <- liftIO Chronos.now
+  withFrozenCallStack (logMsg Message { location = callStack, .. })
+
+fmtMessage :: Message -> Text
+fmtMessage Message {..} = showScope scope <> showSeverity severity location <> showTime time <> message
+
+logMessageStdout :: MonadIO m => LogAction m Message
+logMessageStdout = fmtMessage >$< logTextStdout
+
+logMessageStderr :: MonadIO m => LogAction m Message
+logMessageStderr = fmtMessage >$< logTextStderr
+
+logMessageStdoutAndStderr :: MonadIO m => LogAction m Message
+logMessageStdoutAndStderr = logOut <> logErr
+ where
+  logErr = cfilter (\Message {..} -> severity == Error || severity == Panic) logMessageStderr
+  logOut = cfilter (\Message {..} -> severity /= Error && severity /= Panic) logMessageStdout
+
+showScope :: Scope -> Text
+showScope scope = color Cyan . square <| text
+ where
+  text = case scope of
+    Cli -> "Cli"
+    Api -> "Api"
+
+showTime :: Time -> Text
+showTime t = square . toStrict . TB.toLazyText <| builderDmyHMSz (Chronos.timeToDatetime t)
+
+builderDmyHMSz :: Chronos.Datetime -> TB.Builder
+builderDmyHMSz (Chronos.Datetime date time) =
+  builderDmy date <> spaceSep <> Chronos.builder_HMS (Chronos.SubsecondPrecisionFixed 3) (Just ':') time <> spaceSep <> Chronos.builderOffset
+    Chronos.OffsetFormatColonOn
+    (Chronos.Offset 0)
+ where
+  spaceSep :: TB.Builder
+  spaceSep = TB.singleton ' '
+
+  builderDmy :: Chronos.Date -> TB.Builder
+  builderDmy (Chronos.Date (Chronos.Year y) m d) =
+    zeroPadDayOfMonth d <> spaceSep <> TB.fromText (Chronos.caseMonth Chronos.abbreviated m) <> spaceSep <> TB.decimal y
+
+  zeroPadDayOfMonth :: Chronos.DayOfMonth -> TB.Builder
+  zeroPadDayOfMonth (Chronos.DayOfMonth d) = if d < 100 then Vector.unsafeIndex twoDigitTextBuilder d else TB.decimal d
+
+  twoDigitTextBuilder :: Vector.Vector TB.Builder
+  twoDigitTextBuilder = Vector.fromList . map (TB.fromText . Text.pack) <| twoDigitStrings
+  {-# NOINLINE twoDigitTextBuilder #-}
+
+  twoDigitStrings :: [String]
+  twoDigitStrings = (\a b -> mappend (show @_ @Int a) (show @_ @Int b)) <$> [0 .. 9] <*> [0 .. 9]
+
+showSourceLoc :: CallStack -> Text
+showSourceLoc cs = case getCallStack cs of
+  [] -> "<unknown loc>"
+  [(name, loc)] -> showLoc name loc
+  (_, loc) : (callerName, _) : _ -> showLoc callerName loc
+  where showLoc name SrcLoc {..} = Text.pack srcLocModule <> "." <> Text.pack name <> "#" <> Text.pack (show srcLocStartLine)
+
+showSeverity :: Severity -> CallStack -> Text
+showSeverity Debug   _        = color Green . square <| "Debug"
+showSeverity Info    _        = color Blue . square <| "Info"
+showSeverity Warning _        = color Yellow . square <| "Warning"
+showSeverity Error   _        = color Red . square <| "Error"
+showSeverity Panic   location = color Red . square <| "Panic " <> showSourceLoc location
```

Comment:
This function belongs to the `Cascade.Log.Severity` module. Also, it is pretty-printing the `Severity` datatype so as the convention we can call it `prettyPrintSeverity`.